crates/uv-auth/src/middleware.rs

```diff
@@ -100,88 +128,184 @@ impl Middleware for AuthMiddleware {
+                ))
+            })?;
+
+            let response = next.clone().run(request, extensions).await?;
+
+            // And on failure, search for credentials
+            if matches!(
+                response.status(),
+                StatusCode::FORBIDDEN | StatusCode::NOT_FOUND | StatusCode::UNAUTHORIZED
+            ) {
+                trace!(
+                    "Request for {url} failed with {}, checking for credentials",
+                    response.status()
+                );
+                let credentials = self.cache().get_realm(
+                    NetLoc::from(retry_request.url()),
+                    credentials
+                        .map(|credentials| credentials.to_username())
+                        .unwrap_or(Username::none()),
+                );
+
+                if let Some(credentials) = credentials.as_ref() {
+                    if credentials.password().is_some() {
+                        trace!("Retrying request for {url} with credentials from cache");
+                        retry_request = credentials.authenticate(retry_request);
+                        return self
+                            .complete_request(None, retry_request, extensions, next)
+                            .await;
+                    }
+                }
+
+                if let Some(credentials) = self
+                    .fetch_credentials(credentials.as_deref(), retry_request.url())
+                    .await
+                {
+                    retry_request = credentials.authenticate(retry_request);
+                    trace!("Retrying request for {url} with {credentials:?}");
+                    self.complete_request(
+                        Some(Arc::new(credentials)),
+                        retry_request,
+                        extensions,
+                        next,
+                    )
+                    .await
+                } else {
+                    Ok(response)
+                }
+            } else {
+                Ok(response)
+            }
+        }
+    }
+}
+
+impl AuthMiddleware {
+    /// Run a request to completion.
+    ///
+    /// If credentials are present, insert them into the cache on success.
+    async fn complete_request(
+        &self,
+        credentials: Option<Arc<Credentials>>,
+        request: Request,
+        extensions: &mut Extensions,
+        next: Next<'_>,
+    ) -> reqwest_middleware::Result<Response> {
+        if let Some(credentials) = credentials {
+            let url = request.url().clone();
+            let result = next.run(request, extensions).await;
+
+            // Update the cache with new credentials on a successful request
+            if result
+                .as_ref()
+                .is_ok_and(|response| response.error_for_status_ref().is_ok())
+            {
+                trace!("Updating cached credentials for {url} to {credentials:?}");
+                self.cache().insert(&url, credentials)
```

Comment:
So does every URL that we visit get inserted into the cache?

Reply:
Why do we need to track on a per-URL basis?

Reply:
It allows us to cache credentials more precisely than the realm, so we can handle realms with mixed credentials.